lib/Backend/GlobOptArrays.cpp

```diff
@@ -0,0 +1,1896 @@
+                baseValueType.GetObjectType() == ObjectType::Float64VirtualArray) &&
+                !instr->IsDstNotAlwaysConvertedToNumber()
+                )
+            )
+        {
+            // Unless we're in asm.js (where it is guaranteed that virtual typed array accesses cannot read/write beyond 4GB),
+            // check the range of the index to make sure we won't access beyond the reserved memory beforing eliminating bounds
+            // checks in jitted code.
+            if (!globOpt->GetIsAsmJSFunc() && baseOwnerIndir)
+            {
+                IR::RegOpnd * idxOpnd = baseOwnerIndir->GetIndexOpnd();
+                if (idxOpnd)
+                {
+                    StackSym * idxSym = idxOpnd->m_sym->IsTypeSpec() ? idxOpnd->m_sym->GetVarEquivSym(nullptr) : idxOpnd->m_sym;
+                    Value * idxValue = globOpt->CurrentBlockData()->FindValue(idxSym);
+                    IntConstantBounds idxConstantBounds;
+                    if (idxValue && idxValue->GetValueInfo()->TryGetIntConstantBounds(&idxConstantBounds))
+                    {
+                        BYTE indirScale = Lowerer::GetArrayIndirScale(baseValueType);
+                        int32 upperBound = idxConstantBounds.UpperBound();
+                        int32 lowerBound = idxConstantBounds.LowerBound();
+                        if (lowerBound >= 0 && ((static_cast<uint64>(upperBound) << indirScale) < MAX_ASMJS_ARRAYBUFFER_LENGTH))
+                        {
+                            eliminatedLowerBoundCheck = true;
+                            eliminatedUpperBoundCheck = true;
+                            canBailOutOnArrayAccessHelperCall = false;
+                        }
+                    }
+                }
+            }
+            else
+            {
+                if (!baseOwnerIndir)
+                {
+                    Assert(instr->m_opcode == Js::OpCode::InlineArrayPush ||
+                        instr->m_opcode == Js::OpCode::InlineArrayPop ||
+                        instr->m_opcode == Js::OpCode::LdLen_A);
+                }
+
+                eliminatedLowerBoundCheck = true;
+                eliminatedUpperBoundCheck = true;
+                canBailOutOnArrayAccessHelperCall = false;
+            }
+        }
+    }
+#endif
+}
+
+void GlobOpt::ArraySrcOpt::TryEleminiteBoundsCheck()
+{
+    AnalysisAssert(baseOwnerIndir);
+    Assert(needsHeadSegmentLength);
+
+    // Bound checks can be separated from the instruction only if it can bail out instead of making a helper call when a
+    // bound check fails. And only if it would bail out, can we use a bound check to eliminate redundant bound checks later
+    // on that path.
+    doExtractBoundChecks = (headSegmentLengthIsAvailable || doHeadSegmentLengthLoad) && canBailOutOnArrayAccessHelperCall;
+
+    // Get the index value
+    IR::RegOpnd * const indexOpnd = baseOwnerIndir->GetIndexOpnd();
```

Comment:
nit: Was there a reason that we had a do while(false) here before? can you check the history to see if that was added later.

Reply:
My understanding of reading the original code was that it was done as a do/while so `break` could be used to exit the block. As a separate function return can be done instead. I'll check the history to be sue.